cpp/include/cuspatial/experimental/detail/point_polygon_distance.cuh

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/detail/iterator.hpp>
+#include <cuspatial/detail/utility/device_atomics.cuh>
+#include <cuspatial/detail/utility/linestring.cuh>
+#include <cuspatial/detail/utility/offset_to_keys.cuh>
+#include <cuspatial/detail/utility/zero_data.cuh>
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/detail/algorithm/is_point_in_polygon.cuh>
+#include <cuspatial/experimental/ranges/range.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/tabulate.h>
+
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief For each point in the multipoint, compute point-in-multipolygon in corresponding pair.
+ */
+template <typename MultiPointRange, typename MultiPolygonRange>
+struct point_in_multipolygon_test_functor {
+  MultiPointRange multipoints;
+  MultiPolygonRange multipolygons;
+
+  point_in_multipolygon_test_functor(MultiPointRange multipoints, MultiPolygonRange multipolygons)
+    : multipoints(multipoints), multipolygons(multipolygons)
+  {
+  }
+
+  template <typename IndexType>
+  uint8_t __device__ operator()(IndexType pidx)
+  {
+    auto point        = thrust::raw_reference_cast(multipoints.point(pidx));
+    auto geometry_idx = multipoints.geometry_idx_from_point_idx(pidx);
+
+    bool intersects = false;
+    for (auto polygon : multipolygons[geometry_idx]) {
+      intersects = intersects || is_point_in_polygon(point, polygon);
+    }
```

Comment:
Consider this option. `any_of` makes the intent clearer than a raw loop.

```suggestion
    auto const& polys = multipolygons[geometry_idx];
    bool intersects = thrust::any_of(thrust::seq, polys.begin(), polys.end(), [] (auto poly) { 
        return is_point_in_polygon(point, poly); 
    });
```

Reply:
Not sure about performance, though.  any_of can early exit, at least. (You could add early exit to your loop, too.)

Reply:
> Not sure about performance, though.
I'll write benchmark after this PR. I like readability of `any_of`.